bokeh/bokeh
# Proof of concept: client-side SVG generation

Experimental PR to investigate generating SVGs on the client side instead of 

borrows from: #3867

- [x] issues: fixes #538 
- [x] release document entry (if new feature or API change)

What doesn't currently work:
* ~~SaveTool (should be easy to fix, just needs to use `getSerializedSvg` method instead of `getDataURI` on mock canvas element)~~ fixed
* Glyph clipping needs some tweaking, they are able to be dragged outside of the frame
* There's some artifact-ing from glyphs that are dragged off the canvas, leaving some lines on the canvas edge
* Performance (when panning, which causes lots of rerenders) isn't as fast as canvas. Unclear if is limitation of svg output.
* ~~Exporting layouts. This makes each plot a unique SVG, so it'd require some work to grab all of the SVGs then stitch them together.~~ out-of-scope for PR
* ~~It's not headless. I'd waiting until the ``export`` PR is merged then build on top of that to support headless generation using PhantomJS/headless Chrome. I also haven't experimented with node-canvas to see what the issues are with using that approach recently, though I suspect there are several.~~ out-of-scope for PR

Things that won't work with this approach:
* Exporting plot with a toolbar (the tool icons are overlaid HTML elements, so the exported SVG won't contain the toolbar)
* Exporting plot with webgl glyphs (the webgl glyphs are contained in a overlaid webgl layer, this also won't be exported along with the svg)

Usage/Test suggestions:
* set `Plot.background_fill_color = None` and `Plot.border_fill_color` so that the plot whitespace is transparent and not filled white.
* set `Plot.toolbar_location = None`, so that the toolbar (that isn't exported) doesn't figure into the layout solver (making the plot unnecessarily small to accomodate). But then you can't use the save tool... I need to think about this. Alternatively, you could right-click and save element (at least in FireFox)
